### PR TITLE
UI

### PR DESCRIPTION
**Pull request by tetuaoro**
### Setup page
- [x] Choose lang
- [x] Create first admin
- [x] Set the host and port for client to connect to the server
- [x] If need to restore old clients, restore t

### Login page 
- [x] Login page

### Admin page
- [ ] Setting page *(me.vue)*
- [ ] Dashboard page

### Optional
- Track setup in server side

### CODE REVIEW

**Comment by kaaax0815** [comment]
Setup/Migration
Setup/Wireguard (Host, Port)
still missing. If you want I can make the backend logic and a simple ui for it and you can make it better.

Im a little busy rn tho

**Comment by tetuaoro** [comment]
> Setup/Migration Setup/Wireguard (Host, Port) still missing.

Oh, ok !

**Comment by kaaax0815** [comment]
Also there were quite some hydration issues

**Comment by kaaax0815** [comment]
If you look in the Developer console you probably have some hydration mismatches

**Comment by kaaax0815** [comment]
We have to edit the `localeDetector.ts` as it uses the browser language to try to guess the locale. even tho it should use the database for that.
Or we allow the user to select any language and don't make it fixed

File: src/app/pages/setup.vue [hunk 1]
```diff
@@ -89,28 +112,36 @@ import { FetchError } from 'ofetch';
...
+    value: 'de',
+    name: 'Deutsch',
+  },
+  {
+    value: 'en',
+    name: 'English',
+  },
+  {
+    value: 'de',
+    name: 'Français',
```
Comment (kaaax0815) [inline src/app/pages/setup.vue:140]: should probably be `fr`

Comment (tetuaoro) [reply src/app/pages/setup.vue:140]: yeah thx

**Comment by tetuaoro** [comment]
@kaaax0815 need help with hydration mismatches

### step to reproduce
1. delete container
2. run new container with `docker compose -f *file* up`
3. enter setup page
4. choose a lang (chinese lang for example xD)
5. move toward step
6. reload page

Hydration error occurs

File: src/app/pages/setup.vue [hunk 2]
```diff
@@ -1,106 +1,206 @@
...
-          }}</small>
+    <UiBanner />
+    <Panel>
+      <PanelBody class="md:w-[70%] lg:w-[60%] mx-auto mt-10 p-4">
+        <h2 class="mt-8 mb-16 text-3xl font-medium">
+          {{ $t('setup.welcome') }}
+        </h2>
+
+        <div v-if="step === 1">
+          <p class="text-lg p-8">{{ $t('setup.msgStepOne') }}</p>
```
Comment (kaaax0815) [inline src/app/pages/setup.vue:11]: use `text-center` more

**Comment by kaaax0815** [comment]
Can't test
```
wg-easy-1  |  ERROR  Vue app aliases are not allowed in server runtime. [importing #build/i18n.options.mjs from node_modules/.pnpm/@nuxtjs+i18n@8.5.1_magicast@0.3.5_rollup@4.21.2_vue@3.4.38_typescript@5.5.4_/node_modules/@nuxtjs/i18n/dist/runtime/composables/index.js]
wg-easy-1  | 
wg-easy-1  | 
wg-easy-1  | [nitro]  ERROR  RollupError: node_modules/.pnpm/@nuxtjs+i18n@8.5.1_magicast@0.3.5_rollup@4.21.2_vue@3.4.38_typescript@5.5.4_/node_modules/@nuxtjs/i18n/dist/runtime/routing/utils.js (2:9): "isRef" is not exported by "virtual:#imports", imported by "node_modules/.pnpm/@nuxtjs+i18n@8.5.1_magicast@0.3.5_rollup@4.21.2_vue@3.4.38_typescript@5.5.4_/node_modules/@nuxtjs/i18n/dist/runtime/routing/utils.js".
wg-easy-1  | 
wg-easy-1  | 
wg-easy-1  | 1: import { isString, isSymbol, isFunction } from "@intlify/shared";
wg-easy-1  | 2: import { isRef, unref } from "#imports";
wg-easy-1  |             ^
wg-easy-1  | 3: export const inBrowser = typeof window !== "undefined";
wg-easy-1  | 4: export function getNormalizedLocales(locales) {
```

File: src/app/components/ui/ChooseLang.vue [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            </SelectItemText>
+          </SelectItem>
+        </SelectViewport>
+      </SelectContent>
+    </SelectPortal>
+  </SelectRoot>
+</template>
+
+<script setup lang="ts">
+import { LOCALES } from '~/../i18n.config.js';
```
Comment (kaaax0815) [inline src/app/components/ui/ChooseLang.vue:34]: remove file extension

File: src/server/utils/types.ts [hunk 4]
```diff
@@ -1,5 +1,9 @@
 import type { ZodSchema } from 'zod';
 import { z, ZodError } from 'zod';
+import type { H3Event, EventHandlerRequest } from 'h3';
+import { useI18n } from '#i18n';
+
+const { availableLocales } = useI18n();
```
Comment (kaaax0815) [inline src/server/utils/types.ts:6]: probably just use exported `LOCALES`

**Review by kaaax0815 — CHANGES_REQUESTED** [review]

**Comment by tetuaoro** [comment]
Hi man, how are you today ? I'm on the host/port configuration section today. I'd just like to say that I've run into a problem with the i18n and I don't know why.

When the key/value doesn't exist in another language, the default value (en) seems not to be used by i18n.

**Comment by kaaax0815** [comment]
> When the key/value doesn't exist in another language, the default value (en) seems not to be used by i18n.

You have to set the fallback language. https://i18n.nuxtjs.org/docs/guide/locale-fallback

**Comment by tetuaoro** [comment]
```bash
wg-easy-1  | 2024-09-29T16:54:21.275Z LowDB GET Clients
wg-easy-1  | 2024-09-29T16:54:21.276Z WireGuard Saving Config...
wg-easy-1  | 2024-09-29T16:54:21.276Z WireGuard Config saved successfully.
wg-easy-1  | $ wg-quick down wg0
wg-easy-1  | [nitro] dev:reload: 616.669ms
wg-easy-1  | $ wg-quick up wg0
wg-easy-1  | 
wg-easy-1  |  ERROR  [nitro] [unhandledRejection] Command failed: wg-quick up wg0
wg-easy-1  | [#] 
wg-easy-1  | [#] ip link add wg0 type wireguard
wg-easy-1  | [#] wg setconf wg0 /dev/fd/63
wg-easy-1  | [#] ip -4 address add 10.8.0.1/24 dev wg0
wg-easy-1  | [#] ip -6 address add fdcc:ad94:bacf:61a4::cafe:1/112 dev wg0
wg-easy-1  | [#] ip link set mtu 1420 up dev wg0
wg-easy-1  | [#] iptables -t nat -A POSTROUTING -s 10.8.0.0/24 -o eth0 -j MASQUERADE; iptables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT; iptables -A FORWARD -i wg0 -j ACCEPT; iptables -A FORWARD -o wg0 -j ACCEPT; ip6tables -t nat -A POSTROUTING -s fdcc:ad94:bacf:61a4::cafe:0/112 -o eth0 -j MASQUERADE; ip6tables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT; ip6tables -A FORWARD -i wg0 -j ACCEPT; ip6tables -A FORWARD -o wg0 -j ACCEPT;
wg-easy-1  | modprobe: can't load module ip6_tables (kernel/net/ipv6/netfilter/ip6_tables.ko.zst): invalid module format
wg-easy-1  | modprobe: can't load module ip6_tables (kernel/net/ipv6/netfilter/ip6_tables.ko.zst): invalid module format
wg-easy-1  | ip6tables v1.8.10 (legacy): can't initialize ip6tables table nat': Table does not exist (do you need to insmod?)
wg-easy-1  | Perhaps ip6tables or your kernel needs to be upgraded.
wg-easy-1  | [#] ip link delete dev wg0
wg-easy-1  | 
wg-easy-1  | 
wg-easy-1  |   [#]
wg-easy-1  |   [#] ip link add wg0 type wireguard
wg-easy-1  |   [#] wg setconf wg0 /dev/fd/63
wg-easy-1  |   [#] ip -4 address add 10.8.0.1/24 dev wg0
wg-easy-1  |   [#] ip -6 address add fdcc:ad94:bacf:61a4::cafe:1/112 dev wg0
wg-easy-1  |   [#] ip link set mtu 1420 up dev wg0
wg-easy-1  |   [#] iptables -t nat -A POSTROUTING -s 10.8.0.0/24 -o eth0 -j MASQUERADE; iptables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT; iptables -A FORWARD -i wg0 -j ACCEPT; iptables -A FORWARD -o wg0 -j ACCEPT; ip6tables -t nat -A POSTROUTING -s fdcc:ad94:bacf:61a4::cafe:0/112 -o eth0 -j MASQUERADE; ip6tables -A INPUT -p udp -m udp --dport 51820 -j ACCEPT; ip6tables -A FORWARD -i wg0 -j ACCEPT; ip6tables -A FORWARD -o wg0 -j ACCEPT;
wg-easy-1  |   modprobe: can't load module ip6_tables (kernel/net/ipv6/netfilter/ip6_tables.ko.zst): invalid module format
wg-easy-1  |   modprobe: can't load module ip6_tables (kernel/net/ipv6/netfilter/ip6_tables.ko.zst): invalid module format
wg-easy-1  |   ip6tables v1.8.10 (legacy): can't initialize ip6tables table nat': Table does not exist (do you need to insmod?)
wg-easy-1  |   Perhaps ip6tables or your kernel needs to be upgraded.
wg-easy-1  |   [#] ip link delete dev wg0
wg-easy-1  |   
wg-easy-1  |   at genericNodeError (node:internal/errors:984:15)
wg-easy-1  |   at wrappedFn (node:internal/errors:538:14)
wg-easy-1  |   at ChildProcess.exithandler (node:child_process:422:12)
wg-easy-1  |   at ChildProcess.emit (node:events:519:28)
wg-easy-1  |   at maybeClose (node:internal/child_process:1105:16)
wg-easy-1  |   at ChildProcess._handle.onexit (node:internal/child_process:305:5)
wg-easy-1  | 
wg-easy-1  | [nitro-runtime] error: 0.132ms
```

**Comment by kaaax0815** [comment]
It looks like your kernel modules are somehow broken.

Try
```shell
modprobe ip6_tables
```

Or edit the database and remove iptable scripts as they are not needed for ui development

**Comment by kaaax0815** [comment]
@tetuaoro what's your status on this?

**Comment by tetuaoro** [comment]
> @tetuaoro what's your status on this?

Hi. For now, I can only contribute the week-end.

**Comment by kaaax0815** [comment]
> Hi. For now, I can only contribute the week-end.

No problem! Thank you for your help. Highly appreciated.

What in your opinion has to be done still for this pr to be merged?

**Review by tetuaoro — COMMENTED** [review]
Done

**Comment by tetuaoro** [comment]
Hi,
I have a question : is a customer also a user ? Do we allow the client (user) to connect to the web interface to retrieve his qrcode himself ?

**Comment by kaaax0815** [comment]
@tetuaoro For the migration you could use some of the code I had written here https://github.com/kaaax0815/wg-easy/blob/7a6f7db4d3560f29fb8246b7e6471dfa73edcb30/src/server/utils/config.ts#L12. Maybe users also should be able to pass their old env vars to the ui. my goal is it that users don't have to give out new configs

**Comment by kaaax0815** [comment]
> Hi, I have a question : is a client also a user ? Do we allow the client (user) to connect to the web interface to retrieve his qrcode himself ?

no as then you would have to set a password for each peer. and there is unfortunately no way of using the config to authenticate (which would be sick but not possible with own implementations of WireGuard, etc).

The admin can create users in the ui.
Assign groups to users
And assign clients to groups

This is what I thought makes most sense. But I don't know exactly what the users want. Maybe you have a better idea in mind?

**Comment by kaaax0815** [comment]
And as you already mentioned in the description checking if a user exists to see if the setup is done is not good.

We should add maybe a setup key to the db. which increments for each step. And also split each step into its own page /setup/1, /setup/2, ... or /setup/general, /setup/user, /setup/migration, .... And then depending on the setup key redirect to the correct page (this would allow for adding more setup steps after the fact; even if user already has a working setup)

This is what I though will work, but maybe you have a better idea

**Comment by kaaax0815** [comment]
The Port in the HostPort Section should be supplied by default.
The Text in the Welcome Page should be adjusted the also target new users
The Migration step should be optional or skippable

**Comment by kaaax0815** [comment]
I am currently short on time due to another project, but I will be able to contribute more significantly in the coming weeks.

**Review by kaaax0815 — APPROVED** [review]

**Comment by kaaax0815** [comment]
Merging so I can work on it

**Pull request merged**